test/unit/key_def.cc

```diff
@@ -457,16 +460,1330 @@ test_tuple_validate_key_parts_raw(void)
+
+		tuples_gt.push_back(test_tuple_new("[%u%uNILNIL]", 0, 1));
+		tuples_gt.push_back(test_tuple_new("[%u%u%uNIL]", 0, 0, 1));
+
+		tuples_gt.push_back(test_tuple_new("[%u%u%uNIL]", 0, 1, 0));
+		tuples_gt.push_back(test_tuple_new("[%u%u%u%u]", 0, 0, 1, 1));
+
+		tuples_gt.push_back(test_tuple_new("[%u%u%u]", 0, 1, 0));
+		tuples_gt.push_back(test_tuple_new("[%u%u%u%u]", 0, 0, 1, 1));
+	}
+
+	fail_unless(tuples_eq.size() % 2 == 0);
+	for (size_t i = 0; i < tuples_eq.size(); i += 2) {
+		struct tuple *a = tuples_eq[i];
+		struct tuple *b = tuples_eq[i + 1];
+		test_tuple_compare_with_key(a, b, key_def, 0, funcname);
+		test_tuple_compare_with_key(b, a, key_def, 0, funcname);
+	}
+
+	fail_unless(tuples_gt.size() % 2 == 0);
+	for (size_t i = 0; i < tuples_gt.size(); i += 2) {
+		struct tuple *a = tuples_gt[i];
+		struct tuple *b = tuples_gt[i + 1];
+		test_tuple_compare_with_key(a, b, key_def, 1, funcname);
+		test_tuple_compare_with_key(b, a, key_def, -1, funcname);
+	}
+
+	for (size_t i = 0; i < tuples_eq.size(); i++) {
+		tuple_delete(tuples_eq[i]);
+	}
+
+	for (size_t i = 0; i < tuples_gt.size(); i++) {
+		tuple_delete(tuples_gt[i]);
+	}
+
+	key_def_delete(key_def);
+
+	free((void *)funcname);
+
+	footer();
+	check_plan();
+}
+
+static void
+test_tuple_compare(struct tuple *tuple_a, struct tuple *tuple_b,
+		   struct key_def *cmp_def, int expected,
+		   const char *funcname)
+{
+	int rc = tuple_compare(tuple_a, HINT_NONE,
+			       tuple_b, HINT_NONE,
+			       cmp_def);
+	int rc_hint = tuple_compare(tuple_a, tuple_hint(tuple_a, cmp_def),
+				    tuple_b, tuple_hint(tuple_b, cmp_def),
+				    cmp_def);
+	ok(rc == expected, "%s(%s, %s) = %d, expected %d.", funcname,
+	   tuple_str(tuple_a), tuple_str(tuple_b), rc, expected);
+	fail_unless(rc == rc_hint); /* The fail cond is printed above. */
+}
+
+static void
+test_tuple_compare_slowpath(bool is_nullable, bool has_optional_parts,
+			    bool is_unique)
+{
+	size_t p = 12 + (is_nullable ? 10 : 0) + (has_optional_parts ? 50 : 0);
+	plan(p);
+	header();
+
+	/* has_optional_parts is only valid if is_nullable. */
+	fail_unless(!has_optional_parts || is_nullable);
+
+	struct key_def *pk_def = test_key_def_new(
+		"[{%s%u%s%s}]",
+		"field", 0, "type", "unsigned");
+
+	/* Type is number to prevent using precompiled comparators. */
+	const bool last_is_nullable = has_optional_parts;
+	struct key_def *key_def = test_key_def_new(
+		"[{%s%u%s%s}{%s%u%s%s%s%b}{%s%u%s%s%s%b}]",
+		"field", 1, "type", "number",
+		"field", 2, "type", "number", "is_nullable", is_nullable,
+		"field", 3, "type", "number", "is_nullable", last_is_nullable);
+
+	struct key_def *cmp_def = key_def_merge(key_def, pk_def);
+
+	if (is_unique) {
+		/*
+		 * It's assumed that PK and SK index different parts. So we
+		 * cover cmp_def->unique_part_count < cmp_def->part_count
+		 * branch of the slowpath comparator (its last loop).
+		 */
+		fail_unless(cmp_def->unique_part_count > key_def->part_count);
```

Comment:
It seems to me that this check should be outside the `if` closure.

Reply:
Moved.